A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,184 @@
+additional policy, `LEAST_REQUEST`.
+
+The configuration for the Least Request LB policy is the
+[`least_request_lb_config`
+field](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L916).
+The field is optional; if not present, defaults will be assumed for all
+of its values.  gRPC will support the
+[`choice_count`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L346)
+field the same way that Envoy does.
+The
+[`active_request_bias`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/cluster/v3/cluster.proto#L371)
+field will be ignored as it is only used in the "*all weights not equal*" (weighted least request)
+case which will not be supported as part of this proposal.
+
+#### Changes to the `xds_cluster_resolver` Policy
+
+As a result of
+[gRFC A42: xDS Ring Hash LB Policy](https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md),
+the `xds_cluster_resolver` policy implementation will need additional logic to understand the setup for `LEAST_REQUEST`
+in addition to `ROUND_ROBIN` and `RING_HASH`.
+For `LEAST_REQUEST`, the policy name will be `"LEAST_REQUEST"`, and the config will be the one for the `least_request_experimental` LB Policy described below.
+When the `xds_cluster_resolver` policy sees `"LEAST_REQUEST"`, it will also assume that the
+locality-picking policy is `weighted_target` and the endpoint-picking policy is `least_request_experimental`.
+
+#### `least_request_experimental` LB Policy
+
+We will implement a `least_request_experimental` LB policy that uses the same
+algorithm as Envoy's implementation.  However, this policy will be
+implemented in a non-xDS-specific way, so that it can also be used without
+xDS in the future.
+
+##### LB Policy Config
+
+The `least_request_experimental` policy will have the following config:
+
+```proto
+message LeastRequestLoadBalancingConfig {
+  uint32 choice_count = 1;  // Optional, defaults to 2.
+}
+```
+
+The `choice_count` determines the number of randomly sampled subchannels which should be compared
+for each pick by the picker.
+
+##### Subchannel state handling
+
+The subchannel state handling implemented in the `round_robin` policy will be re-used for `least_request_experimental`.
+Implementations may break out this subchannel state handling and have `round_robin` and `least_request_experimental`
+inherit the logic instead. This would be similar to the "base balancer" already implemented in grpc-go
+[here](https://github.com/grpc/grpc-go/blob/03268c8ed29e801944a2265a82f240f7c0e1b1c3/balancer/base/balancer.go).
+
+TODO: Possibly flesh this out with more explicit logic pending proposal PR review.
+
+##### Aggregated Connectivity State
+
+The `least_request_experimental` policy will use the same heuristic for determining the aggregated
```

Comment:
I don't think we want to go with that approach. I think we'd want to keep round_robin-style. We had to change things up for ring-hash because connection establishment in the degenerate case is serial, which could be _slooow_. For RR and it seems this least-request policy, connecting to backends is done in parallel, so I don't see a need to do anything special.

Reply:
Interesting point! In that case I'll change this to be consistent with RR instead.

Reply:
Updated now. Possibly listing the rules isn't needed here so could remove that if it doesn't make sense.